### ISSUE TITLE
Make website work over https

### ISSUE THREAD

**Issue by zimeon** [IIIF/website#11]
This is still the case, https://iiif.io/ does not work. Current best practice is to make all websites natively https, with http open just to redirect. (Was https://github.com/IIIF/iiif.io/issues/417)

**Comment by glenrobson** [comment]
So due to the impending change with Chrome in 'July' (not sure of the exact date) to require SSL I am looking at changing the website to use SSL.

In the first instance I would like to run the site on both http and https before implementing a redirect. Before I get to that I need a test site that I can play with to find any mixed content issues so I am going to setup preview.iiif.io to be available as http and https. From google it seems the best way to do this is to use CloudFront:

https://medium.com/@sbuckpesch/setup-aws-s3-static-website-hosting-using-ssl-acm-34d41d32e394
and
https://medium.com/@the.benhawy/switching-to-https-for-sites-hosted-on-aws-341a825b0ae5

Issues discovered so far:
 * Seems to be a bucket we no longer use `iiif-preview-website` we now use a bucket called `preview.iiif.io`. I have disabled http sharing on `iiif-preview-website` and will check if anything breaks before removing.
 * the bucket preview.iiif.io currently stores all deployed versions of the site. I need to write something to delete versions which have been merged (i.e. branch no longer exists in github).

**Comment by glenrobson** [comment]
It would also be great to have a test during deploy to pickup any mixed content warnings. I could also use this to get a list of work that needs doing. I've asked on iiif slack.

**Comment by glenrobson** [comment]
Looks useful: https://developers.google.com/web/tools/lighthouse/audits/mixed-content

and also could do a grep for `http:`

**Comment by glenrobson** [comment]
Have preview working over https:

https://preview.iiif.io/root/discovery_list/community/index.html

but doesn't seem to redirect to index.html for:

https://preview.iiif.io/root/discovery_list/community/

you get a 404. Also get cross site issues with showvase.iiif.io

**Comment by glenrobson** [comment]
A very useful post on how to make the default redirect from /about/ forward to /about/index.html:

https://stackoverflow.com/questions/15309113/amazon-cloudfront-doesnt-respect-my-s3-website-buckets-index-html-rules

**Comment by glenrobson** [comment]
Looked into `lighthouse` but from my testing it looked like it only accepted urls rather than working on local html files. 

Also looked at `grep`. It was difficult with `grep` to ignore `http:` in anchor tags as these shouldn't be flagged. If it ignored lines that contained `<a` it missed some images that were on the same line. 

Writing a python script to use [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/) could be an option but I was looking for something ready made. 

In the end I edited @acdha code to create:

https://github.com/IIIF/phantomjs-mixed-content-scan

which works with local html files generated with jekyll. Currently if it finds mixed content and it doesn't stop the build process but this would be good to add in the future. For info the current errors for the website are:

```
Running mixed content report
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/osd.jpg
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/mirador.jpg
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/uv.png
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/leaflet.jpg
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/klokantechiiifviewer.png
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/loris.jpg
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/iip.png
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/georeferencer.jpg
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/yale-ycba-iipmooviewer.png
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/e-codices.png
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/fractals.jpg
❗️  file:///home/travis/build/IIIF/website/_site/test/index.html loaded an insecure resource: http://showcase.iiif.io/img/showcase/satdb.png
… done!
```

and for the api site:

```
Running mixed content report
❗️  file:///home/travis/build/IIIF/api/_site/test/api/annex/cookbook/index.html loaded an insecure resource: http://fonts.googleapis.com/css?family=Ubuntu+Mono
❗️  file:///home/travis/build/IIIF/api/_site/test/api/annex/index.html loaded an insecure resource: http://fonts.googleapis.com/css?family=Raleway:100,300,500
❗️  file:///home/travis/build/IIIF/api/_site/test/api/annex/index.html loaded an insecure resource: http://fonts.googleapis.com/css?family=Open+Sans:400,400italic,600,700
❗️  file:///home/travis/build/IIIF/api/_site/test/api/annex/index.html loaded an insecure resource: http://fonts.googleapis.com/css?family=Inconsolata
❗️  file:///home/travis/build/IIIF/api/_site/test/api/annex/index.html loaded an insecure resource: http://fonts.googleapis.com/css?family=Lato
❗️  file:///home/travis/build/IIIF/api/_site/test/api/image/1.1/change-log/index.html loaded an insecure resource: http://fonts.googleapis.com/css?family=Open+Sans:400,600,700
❗️  file:///home/travis/build/IIIF/api/_site/test/model/shared-canvas/1.0/index.html loaded an insecure resource: http://i.creativecommons.org/l/by-nc-sa/3.0/88x31.png
… done!
```

I think this test will capture most of the mixed content errors but will fail to pick up js scripts calling http locations (this is only a guess I haven't tested) but hopefully if we link to a https version of a js file it would only call https js files...

**Comment by glenrobson** [comment]
Deployed https fixes for the website in https://github.com/IIIF/website/pull/114, next waiting for CloudFront to deploy then can merge:

https://github.com/IIIF/api/pull/1635

Once https://iiif.io is tested I can change the webprefix in _config.yml to point to https://iiif.io. That will mean people switch to the https site if the click the api links from the main website. The api pull already sets the webprefix to https so people will start moving on to the https site once thats deployed. 

The http version will still be accessible for fixtures and context.json links. Need to decided how to proceed with those before automatically forwarding people from http to https.

**Comment by glenrobson** [comment]
CloudFront has deployed and the following is working:

https://iiif.io

**Comment by glenrobson** [comment]
Deployed api pull request and updated the website to default to https. Things that are outstanding:

**API**
 * Compliance link header: http://iiif.io/api/annex/notes/apache/
 * Namespace for iiif URIs: http://iiif.io/api/annex/notes/design_patterns/#use-strings-for-enumerated-flags
 * Example and expanding uri elements: http://iiif.io/api/annex/notes/jsonld/#sample-usage
 * link to iiif context http://iiif.io/api/annex/openannotation/context.json
 * example context + profile: http://iiif.io/api/annex/registry/services/index.html
 * Example contains context: http://iiif.io/api/annex/rest/index.html#extension-context
 * Example contains context and profile http://iiif.io/api/annex/services/index.html
 * IIIF Namespace http://iiif.io/api/auth/0/context.json
 * Context + profile: http://iiif.io/api/auth/0.9/implementation/
 * Context + profiles: http://iiif.io/api/auth/0.9/
 *  IIIF Namespace http://iiif.io/api/auth/1.0/context.json
 * Context + profile: http://iiif.io/api/auth/1.0/implementation/
 * Context + profiles: http://iiif.io/api/auth/1.0/
 * Context http://iiif.io/api/discovery/0.1/index.html
 * Namespace http://iiif.io/api/image/2/context.json
 * Namespace http://iiif.io/api/image/2/info_frame.json
 * Namespace http://iiif.io/api/image/2/level0.json and http://iiif.io/api/image/2/level1.json and http://iiif.io/api/image/2/level2.json
* Namespace http://iiif.io/api/image/2/ontology.xml
 * "The compliance URIs have been moved into the `http://iiif.io/` domain" http://api/image/2.0/change-log.md
 * Profile URIs: http://iiif.io/api/image/2.0/compliance/
 * Context + Profile URIs: http://iiif.io/api/image/2.0/index.html
 * Profile URIs: http://iiif.io/api/image/2.1/compliance.md
 * Context + profile URIs: http://iiif.io/api/image/2.1/index.html
 * Namespace: http://iiif.io/api/image/3/context.json
 * Context + namespace: http://iiif.io/api/image/3/info_frame.json
 * Context + Profiles http://iiif.io/api/image/3.0/index.html
 * Namespace: api/presentation/2/*_frame.json
 * Namespace: http://iiif.io/api/presentation/2/context.json
 * Namespace: http://iiif.io/api/presentation/2/ontology.xml
 * Context + profile: api/presentation/2.0/example/*
 * Context + profiles http://iiif.io//api/presentation/2.0/index.html
 * Context + profiles: http://iiif.io/api/presentation/3/combined-context.json
 * Context + profiles: http://iiif.io/api/presentation/3.0/index.html
 * Namespace: http://iiif.io/api/search/0/context.json
 * Context + profile: http://iiif.io/api/search/0.9/ and `The description block _MUST_ have the `@context` property with the value "http://iiif.io/api/search/{{ page.major }}/context.json", the  `profile` property with the value "http://iiif.io/api/search/{{ page.major }}/search"`
 * Namespace: http://iiif.io/api/search/1/context.json
 * Context + profile: http://iiif.io/api/search/1.0/ 
 * model/shared-canvas/1.0/index.html

So open questions:
 * What do we do with profile URIs that are defined in the spec?
 * Should version 3 include https URIs for context, profile and namspaces?
 * Do we change version 1 & 2?
 * Is it OK if we redirect from http to https for the 2.x and 1.x context.jsons?
 * Do we need 2 context.jsons? One with `@id` of http and one for https?

**Website**
 * Washington conference submissions - propose to leave and will be caught by redirect. 
 * Link in PDF: http://iiif.io/assets/acc_implementation_guide_011017.pdf - propose to leave and will be caught by redirect.

**Comment by glenrobson** [comment]
Issue with redirects sending incorrect host:

"https://www.google.com/search?q=iiif+image+api and click on the first link, the resulting URL in my browser (Chrome) shows up as http://awseb-e-t-awsebloa-2nr6qcm8hsqi-2040100494.us-east-1.elb.amazonaws.com/api/image/2.1/"

**Comment by glenrobson** [comment]
Look at adding CORS header to redirect 301s:

> 
< HTTP/1.1 301 Moved Permanently
< Server: CloudFront
< Date: Wed, 01 Aug 2018 16:13:43 GMT
< Content-Type: text/html
< Content-Length: 183
< Connection: keep-alive
< Location: https://preview.iiif.io/api/https/api/presentation/2/context.json
< X-Cache: Redirect from cloudfront
< Via: 1.1 8425625428d0aaac9b420a3507d8ef76.cloudfront.net (CloudFront)
< X-Amz-Cf-Id: Wuxh3Kl5gKPRiWLNa7aXTaF4eRagHOWEKqtwNCtYgWyqBFgBxgfq9A==
<

**Comment by glenrobson** [comment]
Changes made in https://github.com/IIIF/iiif-reverse-proxy-app and deployed once. Unfortunately it didn't work and lead to circular redirects when looking at the https site. I believe this was due to cloudfront caching both the http and https request path under the same key. I've configured cloudfront to cache based on the following headers:

 * Accept
 * CloudFront-Forwarded-Proto
 * http_x-Forwarded-Proto

which should stop this happening. I will re-deploy the change tonight to see if it now works. The iiif-reverse-proxy-app forwards you to the `https` version of the site if the `x_forwarded_proto` is set to `http` and adds the CORS headers.
